contracts/oracle/CollateralizationOracle.sol

```diff
@@ -0,0 +1,287 @@
+            if (pcvDeposits[i] == _deposit) {
+                found = true;
+                pcvDeposits[i] = pcvDeposits[_nDeposits - 1];
+            }
+        }
+        pcvDeposits.pop();
+
+        // emit event
+        emit DepositRemove(msg.sender, _deposit);
+    }
+
+    /// @notice Set the price feed oracle (in USD) for a given asset.
+    /// @param _token : the asset to add price oracle for
+    /// @param _oracle : price feed oracle for the given asset
+    function setOracle(address _token, address _oracle) external onlyGovernor {
+        // add oracle to the map(ERC20Address) => OracleAddress
+        token2oracle[_token] = _oracle;
+
+        // emit event
+        emit OracleSet(msg.sender, _token, _oracle);
+    }
+
+    // ----------- IOracle override methods -----------
+    /// @notice update all oracles required for this oracle to work
+    function update() external override whenNotPaused {
+        for (uint256 i = 0; i < tokensInPcv.length; i++) {
+            IOracle(token2oracle[tokensInPcv[i]]).update();
+        }
+    }
+
+    // @notice returns true if any of the oracles required for this oracle to
+    //         work are outdated.
+    function isOutdated() external override view returns (bool) {
+        bool _outdated = false;
+        for (uint256 i = 0; i < tokensInPcv.length && !_outdated; i++) {
+            _outdated = _outdated || IOracle(token2oracle[tokensInPcv[i]]).isOutdated();
+        }
+        return _outdated;
+    }
+
+    /// @notice Get the current collateralization ratio of the protocol.
+    /// @return _collateralRatio the current collateral ratio of the protocol.
+    /// @return _globalValid the current oracle validity status (false if any
+    ///         of the oracles for tokens held in the PCV are invalid, or if
+    ///         this contract is paused).
+    function read() public override view returns (Decimal.D256 memory, bool) {
+        // The total amount of FEI controlled (owned) by the protocol
+        uint256 _protocolControlledFei = 0;
+        // The total USD value of assets held in the PCV
+        uint256 _protocolControlledValue = 0;
+        // false if any of the token oracles are invalid, or if paused
+        bool _globalValid = !paused();
+
+        // For each token...
+        for (uint256 i = 0; i < tokensInPcv.length; i++) {
+            address _token = tokensInPcv[i];
+            (Decimal.D256 memory _price, bool _valid) = IOracle(token2oracle[_token]).read();
+            _globalValid = _globalValid && _valid;
+
+            // For each deposit...
+            for (uint256 j = 0; j < token2deposits[_token].length; j++) {
+                IPCVDepositV2 _deposit = IPCVDepositV2(token2deposits[_token][j]);
+
+                // Increment the total protocol controlled value by the USD value of
+                // the asset held in this PCVDeposit
+                _protocolControlledValue += _price.mul(_deposit.resistantBalance()).asUint256();
```

Comment:
Would storing an intermediate value and multiplying by the price once at the end save gas?